### DISCUSSION TITLE
Media Visibility & Page Visibility API

### DISCUSSION THREAD

**Discussion by pietrop** [vidstack/player#691] [Q&A] [answered]
_non urgent question_

Current  [Media Visibility](https://www.vidstack.io/docs/player/react/components/media/media-visibility/)  description in the doc
>This component is used to manage media state as page or viewport visibility changes.

**Question/ 💡:** Should [Media Visibility](https://www.vidstack.io/docs/player/react/components/media/media-visibility/) have some logic to pause the player if the user changes tab? (this could also be an option param with a set default)

[Page Visibility API | MDN](https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API#example)

<details><summary>Example of how I've implemented this in another player in case it helps</summary>

```js
document.addEventListener(getVisibilityChangeNameWithBrowserPrefix(), this.#onPageVisibilityChange.bind(this))   
```
```js
/**
 *
 *
 * @description adds cross browser support for visibility change event
 * @private
 * @see [Page Visibility API - Web APIs | MDN](https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API#example)
 * @return {string}
 * @todo: could be moved to a separate util folder
 */
function getVisibilityChangeNameWithBrowserPrefix() {
	let visibilityChange = "visibilitychange";
	if (typeof document.msHidden !== "undefined") {
		visibilityChange = "msvisibilitychange";
	} else if (typeof document.webkitHidden !== "undefined") {
		visibilityChange = "webkitvisibilitychange";
	}

	return visibilityChange;
}
```
```js
	/**
	 *
	 * @memberof Player
	 * @event MediaPlayer#pagevisibilitychange
	 * @type {event}
	 * @instance
	 */
	#onPageVisibilityChange() {
		if (document.visibilityState === "visible") {
			if (isInViewport(this.#container)) {
				this.#emit(API_EVENTS.PAGE_VISIBILITY_CHANGE, { isVisible: true });
				this.#isInView = true;
			} else {
				this.#isInView = false;
				this.#emit(API_EVENTS.PAGE_VISIBILITY_CHANGE, { isVisible: false });
			}
		} else {
			this.#isInView = false;
			this.#emit(API_EVENTS.PAGE_VISIBILITY_CHANGE, { isVisible: false });
		}
	}
```

</details>

_If it's of interest happy to do a PR for this, with some guidance on where is best to put it_

**Comment by mihar-22** [comment, marked as answer]
I'm thinking we should split `MediaVisibility` up because it's a bit conflated with different responsibilities (i.e., viewport and page). It currently does support [page state and visibility](https://www.vidstack.io/docs/player/react/components/media/media-visibility/#page-state-vs-visibility) setting but I think it's all way too confusing. If you set `pageChangeType="visibility"` it should handle what you're looking for.

The more I read/talk about this component, it's obvious it needs to be refactored. It's pretty much why I labelled it as experimental. If you have any suggestions, happy to listen :)

**Reply by pietrop** [reply]
Makes sense. it could be splitted in 3 components?
- [ ] [page lifecycle state](https://developers.google.com/web/updates/2018/07/page-lifecycle-api#overview_of_page_lifecycle_states_and_events)
- [ ] [document's visibility](https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilityState).
- [ ] Page visibility ([Page Visibility API | MDN](https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API#example)

And there could be a parent component that composes all 3 (eg "Player visibility" - or some better name)for ease of use if one wants to use all 3 at once. Otherwise devs can also cherry pick if they only need one, a different combination of the 3 or to customize the 3 in a different way etc..


Generally these are uses to play/pause the component when it goes out of view. And either keep it paused or resume playing when back in view. Which may or may not create issues if other players are on the page in terms of global orchestration. So there might be a little more to consider in terms of how it works in some of those edge case. But I think if it's setup as building block of functionality then those can be potentially tackled and configured independently. If that makes sense.
